# Тестовое задание для стажеров DS по разработке PD модели

<b>Задание:</b> Необходиом разработать модель предсказания дефолта. Даны финансовые данные 32 395 компаний (выручка, активы, пассивы и т.д). Результат будет оцениваться на скрытой выборке из 200 компаний по метрике качества accuracy_score

<b>Цель задания</b> научиться / показать знания в разработке моделей оценки риска - probability of default. Задание приближено к реальному, данные синтетические. Лидерборд для оценки собственного решения и геймификации процесса обучения. Первое место в лидерборде не явлется целью. Цель - правильно разарботанная модель, удовлетворяющая требованиям банковской системы. В случае возникновения вопросов, сотрудники блока Риски готовы проконсультировать и дать обратную связь на модель (телеграм чат).

<b>Метрика качества</b> accuracy_score 

### Импортируем библиотеки

In [ ]:
!pip install pandas-profiling --upgrade

In [ ]:
import pandas as pd
import numpy as np
import pandas_profiling
from matplotlib import pyplot as plt
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedKFold
from xgboost import XGBClassifier
from sklearn.ensemble import IsolationForest
from sklearn.model_selection import RandomizedSearchCV
from sklearn.calibration import CalibratedClassifierCV

In [ ]:
!pip install skrebate

In [ ]:
from skrebate import ReliefF

### Подгружаем данные

In [ ]:
!wget https://raw.githubusercontent.com/BKHV/ds-test/master/PD-data-train.csv
!wget https://raw.githubusercontent.com/BKHV/ds-test/master/PD-data-test.csv
!wget https://raw.githubusercontent.com/BKHV/ds-test/master/PD-data-desc.csv

--2021-02-09 21:12:01--  https://raw.githubusercontent.com/BKHV/ds-test/master/PD-data-train.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5329268 (5.1M) [text/plain]
Saving to: ‘PD-data-train.csv’

PD-data-train.csv   100%[===================>]   5.08M  20.4MB/s    in 0.2s    

2021-02-09 21:12:03 (20.4 MB/s) - ‘PD-data-train.csv’ saved [5329268/5329268]

--2021-02-09 21:12:03--  https://raw.githubusercontent.com/BKHV/ds-test/master/PD-data-test.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 28259 (28K) [text/plain]
Saving to: ‘

In [ ]:
train_df = pd.read_csv('PD-data-train.csv', sep=';')
test_df = pd.read_csv('PD-data-test.csv', sep=';')
desc_df = pd.read_csv('PD-data-desc.csv', sep=';')

### Описание данных

In [ ]:
desc_df

field  ...                                           desc_rus
0                     ar_revenue  ...   Выручка от реализации (на конец последнего года)
1              ar_total_expenses  ...                     Общие расходы за последний год
2                   ar_sale_cost  ...    Себестоимость продаж (на конец последнего года)
3            ar_selling_expenses  ...    Коммерческие расходы (на конец последнего года)
4         ar_management_expenses  ...  Управленческие расходы (на конец последнего года)
5                 ar_sale_profit  ...   Прибыль от реализации (на конец последнего года)
6   ar_balance_of_rvns_and_expns  ...  Баланс доходов и расходов (на конец последнего...
7           ar_profit_before_tax  ...         Валовая прибыль (на конец последнего года)
8                       ar_taxes  ...  Текущий на налог на прибыль (на конец последне...
9     ar_other_profit_and_losses  ...  Прочие доходы и расходы (на конец последнего г...
10                 ar_net_profit  ...          Чистая прибыль (на конец последнего года)
11         ab_immobilized_assets  ...    Основные внеоборотные средства за последний год
12      ab_mobile_current_assets  ...        Оборотные активы (на конец последнего года)
13                  ab_inventory  ...                  Запасы (на конец последнего года)
14        ab_accounts_receivable  ...  Дебиторская задолженность (на конец последнего...
15       ab_other_current_assets  ...  Прочие оборотные активы (на конец последнего г...
16        ab_cash_and_securities  ...  Денежные средства и денежные эквиваленты (на к...
17                     ab_losses  ...  Нераспределенная прибыль (на конец последнего ...
18                ab_own_capital  ...                 Капитал (на конец последнего года)
19           ab_borrowed_capital  ...  Общие долговые обязательства (на конец последн...
20      ab_long_term_liabilities  ...  Долгосрочный обязательства (на конец последнег...
21       ab_short_term_borrowing  ...  Краткосрочные заемные средства (на конец после...
22           ab_accounts_payable  ...        Кредиторская задодженность за последний год
23           ab_other_borrowings  ...    Прочие обязательства (на конец последнего года)
24                       bus_age  ...                               Срок ведения бизнеса
25                      ogrn_age  ...                   Срок с момента присваивания ОГРН
26                adr_actual_age  ...     Срок с момента регистрации юридического адреса
27               head_actual_age  ...             Срок с момента назначения руководителя
28                cap_actual_age  ...                  Срок с момента установки капитала
29                ul_staff_range  ...                             Количество сотрудников
30                ul_capital_sum  ...                                  Величина капитала
31               ul_founders_cnt  ...                              Количество акционеров
32                 ul_branch_cnt  ...                                Количество филиалов
33              ul_strategic_flg  ...                  Компания стратегического значения
34          ul_systematizing_flg  ...                         Системообразующая компания

[35 rows x 3 columns]

Семплы и размеры выборок

In [ ]:
train_df.head()

record_id  ar_revenue  ...  ul_systematizing_flg  default_12m
0          7  52996000.0  ...                     0            0
1         13         NaN  ...                     0            0
2         18         NaN  ...                     0            0
3         22  32049000.0  ...                     0            0
4         23  16458000.0  ...                     0            0

[5 rows x 37 columns]

In [ ]:
train_df.shape

(32395, 37)

In [ ]:
train_df['default_12m'].value_counts()

0    30304
1     2091
Name: default_12m, dtype: int64

Наблюдается довольно сильный дисбаланс классов

In [ ]:
test_df.head()

record_id   ar_revenue  ...  ul_strategic_flg  ul_systematizing_flg
0        196   39103000.0  ...                 0                     0
1       1196          NaN  ...                 0                     0
2       2813   54174000.0  ...                 0                     0
3       4385    1904000.0  ...                 0                     0
4       6479  225584000.0  ...                 0                     0

[5 rows x 36 columns]

In [ ]:
test_df.shape

(200, 36)

Количество разных типов фичей

In [ ]:
train_df.dtypes.value_counts()

float64    25
int64      11
object      1
dtype: int64

Уровень дефолт-рейта в выборке = 6,7219%

In [ ]:
train_df['default_12m'].values.mean()

0.06454699799351751

In [ ]:
train_df.record_id.nunique()

32395

32395 уникальных наблюдений

### Немного поисследуем данные

Создадим автоматические отчеты через pandas profiling для тестового и тренировочного датасетов. Из-за того, что колаб вылетает после формирования отчета, скачаем их в виде html файлов

In [ ]:
report = pandas_profiling.ProfileReport(train_df)
report.to_file(output_file='train_report.html')

Summarize dataset:   0%|          | 0/50 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [ ]:
report_test = pandas_profiling.ProfileReport(test_df)
report_test.to_file(output_file='test_report.html')

Summarize dataset:   0%|          | 0/49 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

# Мои модели
---

Можно наблюдать, что по фичам с 0 по 23 (без учета record_id) 46.8% пропусков в трейне и 59% пропусков в тесте. При этом пропуски в одних и тех же строчках сразу по всем показателям. Очевидно, пропуски связаны с тем, что для данных компаний просто отсутствуют данные финансовой отчетности. Однако для прогнозирования дефолта данные показатели могут оказаться важны. 

Было решено попробовать разделить данные на 2 части (с пропусками и без) и для каждой из них строить отдельную модель.


---
Избавимся от фич ['ul_branch_cnt', 'ul_strategic_flg', 'ul_systematizing_flg'], т.к. они крайне малоинформативны (можно убедиться в отчете, в трейне они практически константны, в тесте - полностью константы). Также избавимся сразу от 'record_id'. Кроме этого избавимся от ['head_actual_age', 'cap_actual_age'], т.к. они являются полными дубликатами 'adr_actual_age'

In [ ]:
to_drop = ['record_id', 'ul_branch_cnt', 'ul_strategic_flg', 'ul_systematizing_flg', 'head_actual_age', 'cap_actual_age']
train_df.drop(to_drop, axis=1, inplace=True)

Также закодируем категориальную переменную 'ul_staff_range', для начала при помощи простого label-кодирования, но потом можно попробовать что-то поинтереснее (если в этом будет особый смысл)

In [ ]:
train_df['ul_staff_range'].value_counts()

[1-100]      29281
(100-500]     2757
> 500          357
Name: ul_staff_range, dtype: int64

In [ ]:
# Закодируем вручную, чтобы случайно не потерять смысл возрастания количества сотрудников

def simple_staff_coder(x):
    code_dict = {'[1-100]': 0, '(100-500]': 1, '> 500': 2}
    return code_dict[x]

train_df['ul_staff_range'] = train_df['ul_staff_range'].apply(simple_staff_coder)

train_df['ul_staff_range'].value_counts()

0    29281
1     2757
2      357
Name: ul_staff_range, dtype: int64

Теперь разделим данные на 2 части:

In [ ]:
# 0 - cо всеми фичами, 1 - без данных фин. отчетности

X_0 = train_df.dropna()
X_1 = train_df.loc[train_df.iloc[:, :24].isna().any(axis=1), train_df.columns[24:]]

y_0 = X_0['default_12m']
y_1 = X_1['default_12m']

X_0 = X_0.drop('default_12m', axis=1)
X_1 = X_1.drop('default_12m', axis=1)

In [ ]:
# Проверим, не потерялись ли какие-нибудь данные

X_0.shape[0] + X_1.shape[0] == train_df.shape[0]

True

На этом обработка трейна пока закончена. Будем обучаться сразу на всей выборке с кросс-валидацией (без дополнительного train_test_split).

## Генерация дополнительнич фич для 0 группы

13 дополнительных факторов: **nf_1, nf_2, ..., nf_13**

**Факторы из модели дискриминатного анализа Альтмана: (1-5)**
- доля чистого оборотного капитала в сумме активов
- нераспределенная прибыль / активы (рентабельность активов)
- отношение прибыли до уплаты процентов и налогов к величине активов
- отношение собственного капитала к объему заемных средств
- выручка / активы (оборачиваемость активов)

**Факторы из модели Чессера (адаптир. под РФ, без повторяющ., те, что можно рассчитать): (6-9)**
- денежные средства + краткосрочные фин. вложения)/оборотные + внеоборотные активы
- выручка от продаж/денежные средства + краткосрочные фин. вложения)
- долгосрочные обязательства + краткосрочные обязательства)/оборотные + внеоборотные активы)
- оборотный капитал/выручка от продаж

**Показатели для установления неудовлетворительной или удовлетворительной структуры баланса: (10-13)**
- Коэффициент абсолютной ликвидности: денежныесредства + краскосрочные фин. вложения / краткосрочные обязательства
- Коэффициент срочной ликвидности: ден.средства + краскоср.фин.вложения + деб. задолж. / краткосрочные обязательства
- Коэффициент текущей ликвидности: оборотные активы / краткосрочные обязательства
- Коэффициент обеспеченности собственными средствами: собственные оборотные средства / оборотные активы

In [ ]:
# Напишем функцию, чтобы затем легко добавить новые факторы в тестовый датасет

def get_new_features(df_original):
    df = df_original.copy()
    # краткоср. обязательства:
    short_term_lialabilities = df.ab_short_term_borrowing	+ df.ab_accounts_payable + df.ab_other_borrowings 
    # активы:
    assets = df.ab_mobile_current_assets + df.ab_immobilized_assets 
    # денежные средства + краткосрочные фин. вложения:
    # почему-то тут оборотные активы = запасы + дебиторка + прочие, а денежные средства и фин. вложения где???
    # взял здесь поэтому просто денежные средства!!!
    #cash_and_short_investments = df.ab_mobile_current_assets	- (df.ab_inventory + df.ab_accounts_receivable + df.ab_other_current_assets)
    cash_and_short_investments = df.ab_cash_and_securities

    df['nf_1'] = (df.ab_mobile_current_assets - short_term_lialabilities) / assets
    df['nf_2'] = df.ab_losses / assets
    df['nf_3'] = df.ar_profit_before_tax / assets
    df['nf_4'] = df.ab_own_capital / (df.ab_long_term_liabilities + short_term_lialabilities)
    df['nf_5'] = df.ar_revenue / assets
    df['nf_6'] = cash_and_short_investments / assets
    df['nf_7'] = df.ar_revenue / cash_and_short_investments
    df['nf_8'] = (df.ab_long_term_liabilities + short_term_lialabilities) / assets
    df['nf_9'] = (df.ab_mobile_current_assets - short_term_lialabilities) / df.ar_revenue
    df['nf_10'] = cash_and_short_investments / short_term_lialabilities
    df['nf_11'] = (cash_and_short_investments + df.ab_accounts_receivable) / short_term_lialabilities
    df['nf_12'] = df.ab_mobile_current_assets / short_term_lialabilities
    df['nf_13'] = ((df.ab_own_capital + df.ab_long_term_liabilities) - df.ab_immobilized_assets) / df.ab_mobile_current_assets
  
    return df

In [ ]:
# Добавим новые фичи в датасет:

X_0 = get_new_features(X_0)
X_0.head()

ar_revenue  ar_total_expenses  ar_sale_cost  ...     nf_11     nf_12     nf_13
0    52996000.0         48344000.0    48344000.0  ...  0.157951  0.456149  0.091575
3    32049000.0         29568000.0    29568000.0  ...  0.051479  1.162377  0.522733
4    16458000.0         14935000.0    14935000.0  ...  0.653636  6.997654  1.389832
9   125432000.0        122655000.0   122655000.0  ...  1.387089  2.071860  0.456524
10   59570000.0         59331000.0    59331000.0  ...  0.397055  0.988580  0.000263

[5 rows x 43 columns]

Видим, что в 1600 строчках появились пропуски и бесконечности. Это связано с тем, что при расчете показателей знаменатель (или +числитель) обнуляется, т.к. в финансовой отчетности много нулей (в некоторых строчках - все):

In [ ]:
X_0[X_0.replace([np.inf, -np.inf], np.nan).isna().any(axis=1)]

ar_revenue  ar_total_expenses  ...     nf_12      nf_13
14     23142000.0         22064000.0  ...       inf   0.074801
16       860000.0           274000.0  ...  0.000000        NaN
21      4654000.0          4619000.0  ...  1.146552   0.127820
27     13048000.0         13003000.0  ...       inf   0.006980
39       169000.0           113000.0  ...       inf   1.404255
...           ...                ...  ...       ...        ...
32249    787000.0          1272000.0  ...       NaN        NaN
32306   4615000.0          4609000.0  ...       inf   0.224266
32309   2591000.0          1964000.0  ...  0.213930  52.244186
32347   5966000.0          5617000.0  ...  0.833803  -0.199324
32362   1939000.0          1697000.0  ...  0.000000       -inf

[1600 rows x 43 columns]

Было решено такие строки отнести к 1 группе данных (где вообще отсутствуют данные фин. отчетности).

In [ ]:
X_0.replace([np.inf, -np.inf], np.nan, inplace=True)

nans_X_0_bool = X_0.isna().any(axis=1)

to_concat = X_0.loc[nans_X_0_bool, train_df.columns[24:-1]]
to_concat_y = y_0[nans_X_0_bool]

X_1 = pd.concat([X_1, to_concat])
y_1 = pd.concat([y_1, to_concat_y])

y_0 = y_0.drop(X_0[nans_X_0_bool].index)
X_0 = X_0.drop(X_0[nans_X_0_bool].index)

Отфильтруем обе группы данных от выбросов при помощи Isolation Forest

In [ ]:
isol_forest = IsolationForest(n_estimators=250, random_state=0, n_jobs=-1)

isol_forest.fit(X_0)
is_outliers_0 = isol_forest.predict(X_0)
y_0 = y_0.drop(X_0[is_outliers_0==-1].index)
X_0 = X_0.drop(X_0[is_outliers_0==-1].index)

isol_forest.fit(X_1)
is_outliers_1 = isol_forest.predict(X_1)
y_1 = y_1.drop(X_1[is_outliers_1==-1].index)
X_1 = X_1.drop(X_1[is_outliers_1==-1].index)

print(f'''В группе X_0 отфильтровано {(is_outliers_0==-1).sum()} выбросов
В группе X_1 отфильтровано {(is_outliers_1==-1).sum()} выбросов
''')

В группе X_0 отфильтровано 565 выбросов
В группе X_1 отфильтровано 1580 выбросов



## Отбор признаков

Попытаемся для группы X_0 отобрать часть признаков, так как их стало довольно много (43). В X_1 всего 6 признаков, поэтому отбирать лишние особого смысла нет.

In [ ]:
# Сократим количество признаков с 43 до 30 на основе метода ReliefF

n_new_features = 30

feature_filter = ReliefF(n_features_to_select=n_new_features, n_neighbors=100, n_jobs=-1)
X_0_new = feature_filter.fit_transform(X_0.values, y_0.values)

In [ ]:
new_cols = []
cols_to_drop = []

for i in range(n_new_features):
    for col in X_0.columns:
        if (X_0.reset_index().loc[:, col] - pd.DataFrame(X_0_new).iloc[:, i]).sum() == 0:
            new_cols.append(col)

for col in X_0.columns:
    if col not in new_cols:
        cols_to_drop.append(col)

In [ ]:
cols_to_drop

['ar_selling_expenses',
 'ar_management_expenses',
 'ar_balance_of_rvns_and_expns',
 'ar_other_profit_and_losses',
 'ab_other_current_assets',
 'ab_cash_and_securities',
 'ab_other_borrowings',
 'ul_capital_sum',
 'nf_4',
 'nf_9',
 'nf_10',
 'nf_11',
 'nf_13']

In [ ]:
cols_to_drop = [
    'ar_selling_expenses', 'ar_management_expenses', 'ar_balance_of_rvns_and_expns',
    'ar_other_profit_and_losses', 'ab_other_current_assets', 'ab_cash_and_securities', 
    'ab_other_borrowings', 'ul_capital_sum', 'nf_4', 'nf_9', 'nf_10', 'nf_11', 'nf_13'
]

In [ ]:
X_0_new = X_0.drop(cols_to_drop, axis=1)
X_0_new.head()

ar_revenue  ar_total_expenses  ...      nf_8     nf_12
0    52996000.0         48344000.0  ...  2.192265  0.456149
3    32049000.0         29568000.0  ...  0.860306  1.162377
4    16458000.0         14935000.0  ...  0.142905  6.997654
9   125432000.0        122655000.0  ...  0.482658  2.071860
10   59570000.0         59331000.0  ...  1.011551  0.988580

[5 rows x 30 columns]

## Подбор гиперпараметров

In [ ]:
NFOLDS = 5
SEED = 0

param_grid_xgb = {
    'n_estimators': np.arange(150, 325, 25),
    'max_depth': list(np.arange(5, 20)) + [None],
    'skip_drop': np.linspace(0, 0.5, 10),
    'rate_drop': np.linspace(0, 0.5, 10),
    'learning_rate': np.linspace(0, 0.7, 15),
    'subsample': np.linspace(0.7, 1, 7),
    'reg_alpha': np.linspace(0, 1, 11),
    'reg_lambda': np.linspace(0, 1, 11),
}

cv = StratifiedKFold(n_splits=NFOLDS, shuffle=True, random_state=SEED)

clf_xgb_0 = XGBClassifier(random_state=SEED, booster='dart', colsample_bytree=1)
clf_xgb_1 = XGBClassifier(random_state=SEED, booster='dart', colsample_bytree=1)

grid_cv_xgb_0 = RandomizedSearchCV(clf_xgb_0, param_grid_xgb, scoring='balanced_accuracy', cv=cv, n_jobs=-1, n_iter=100, random_state=SEED)
grid_cv_xgb_1 = RandomizedSearchCV(clf_xgb_1, param_grid_xgb, scoring='balanced_accuracy', cv=cv, n_jobs=-1, n_iter=100, random_state=SEED)

In [ ]:
%%time
np.random.seed(SEED)
grid_cv_xgb_0.fit(X_0_new, y_0)

best_params_0 = grid_cv_xgb_0.best_params_

print(f'''XGBoost, группа 0
Оптимальные гиперпараметры:\n{grid_cv_xgb_0.best_params_}\n
Balanced accuracy на кросс-валидации:\n{grid_cv_xgb_0.best_score_}''')

XGBoost, группа 0
Оптимальные гиперпараметры:
{'subsample': 0.75, 'skip_drop': 0.3333333333333333, 'reg_lambda': 0.7000000000000001, 'reg_alpha': 0.6000000000000001, 'rate_drop': 0.3333333333333333, 'n_estimators': 300, 'max_depth': 7, 'learning_rate': 0.7}

Balanced accuracy на кросс-валидации:
0.5875884632617971
CPU times: user 1min 12s, sys: 618 ms, total: 1min 13s
Wall time: 4h 13min 26s


In [ ]:
%%time
np.random.seed(SEED)
grid_cv_xgb_1.fit(X_1, y_1)

best_params_1 = grid_cv_xgb_1.best_params_

print(f'''XGBoost, группа 1
Оптимальные гиперпараметры:\n{grid_cv_xgb_1.best_params_}\n
Balanced accuracy на кросс-валидации:\n{grid_cv_xgb_1.best_score_}''')

XGBoost, группа 1
Оптимальные гиперпараметры:
{'subsample': 0.95, 'skip_drop': 0.4444444444444444, 'reg_lambda': 0.5, 'reg_alpha': 0.30000000000000004, 'rate_drop': 0.3333333333333333, 'n_estimators': 275, 'max_depth': 15, 'learning_rate': 0.49999999999999994}

Balanced accuracy на кросс-валидации:
0.5324217677224543
CPU times: user 1min 9s, sys: 189 ms, total: 1min 10s
Wall time: 37min 50s


### Построение модели на всех данных

In [ ]:
best_params_0 = {
    'subsample': 0.9, 
    'skip_drop': 0.25, 
    'reg_lambda': 0.5, 
    'reg_alpha': 0.3, 
    'rate_drop': 0.15, 
    'n_estimators': 225, 
    'max_depth': 14, 
    'learning_rate': 0.25
}


best_params_1 = {
    'subsample': 0.95, 
    'skip_drop': 0.444, 
    'reg_lambda': 0.5, 
    'reg_alpha': 0.34, 
    'rate_drop': 0.333, 
    'n_estimators': 275, 
    'max_depth': 15, 
    'learning_rate': 0.5
}


In [ ]:
clf_xgb_0.set_params(**best_params_0)
clf_xgb_0.fit(X_0_new.values, y_0);

clf_xgb_0.set_params(**best_params_1)
clf_xgb_1.fit(X_1.values, y_1);

### Формирование файла с ответом

In [ ]:
answ_df = pd.DataFrame()
answ_df['id'] = test_df['record_id']

Для начала проведем на тестовом датасете те же преобразования, что делали в тренировочном

In [ ]:
#test_df = pd.read_csv('PD-data-test.csv', sep=';')

In [ ]:
test_df.drop(to_drop, axis=1, inplace=True)

test_df['ul_staff_range'] = test_df['ul_staff_range'].apply(simple_staff_coder)

X_0_test = test_df.dropna()
X_1_test = test_df.loc[test_df.iloc[:, :24].isna().any(axis=1), test_df.columns[24:]]

X_0_test = get_new_features(X_0_test)

X_0_test.replace([np.inf, -np.inf], np.nan, inplace=True)

nans_X_0_test_bool = X_0_test.isna().any(axis=1)

to_concat_test = X_0_test.loc[nans_X_0_test_bool, test_df.columns[24:]]

X_1_test = pd.concat([X_1_test, to_concat_test])
X_0_test = X_0_test.drop(X_0_test[nans_X_0_test_bool].index)

X_0_test = X_0_test.drop(cols_to_drop, axis=1)

Откалибруем классификатор при помощи изотонической регрессии

In [ ]:
clf_xgb_0_calib = CalibratedClassifierCV(base_estimator=clf_xgb_0, cv='prefit', method='isotonic')
clf_xgb_1_calib = CalibratedClassifierCV(base_estimator=clf_xgb_1, cv='prefit', method='isotonic')

clf_xgb_0_calib.fit(X_0_new.values, y_0);
clf_xgb_1_calib.fit(X_1.values, y_1);

Теперь получим прогнозы и запишем ответ

In [ ]:
cutoff_0 = y_0.mean()
cutoff_1 = y_1.mean() 

preds_0 = (clf_xgb_0_calib.predict_proba(X_0_test.values)[:, 1]>cutoff_0).astype(int)
preds_1 = (clf_xgb_1_calib.predict_proba(X_1_test.values)[:, 1]>cutoff_1).astype(int)

answ_df.loc[X_0_test.index, 'predict'] = preds_0
answ_df.loc[X_1_test.index, 'predict'] = preds_1

In [ ]:
answ_df.to_csv('PD-submit.csv',index=False, sep=';')